### PR TITLE
NF: add fsck check for pseudo-key names in the asset file

### LINKED ISSUE #283: add `fsck` check for pseudo-key names in the asset file
Pseudo Keys (`type`, `make`, `model`, `serial`) are reserved key-names and while they technically /can/ exist in the asset file, they are redundant and will be ignored by `get` and others.

There should be an `fsck` check to ensure that no asset file contains keys with those names.

### PR DESCRIPTION
**Pull request by TobiasKadelka**
Adds the check to `Repo.fsck()`, and adds tests.

Close #283

### CODE REVIEW

File: onyo/lib/onyo.py [hunk 1]
```diff
@@ -435,13 +435,15 @@ def fsck(self, tests: Optional[list[str]] = None) -> None:
         - "asset-yaml": loads each assets and checks if it's valid YAML
         - "asset-validity": loads each asset and validates the contents against
           the validation rulesets defined in ``.onyo/validation/``.
+        - "no-pseudo-keys": verifies that assets do not contain pseudo-key names
```
Comment (aqw) [inline onyo/lib/onyo.py:438]: IMO, the `no` should be dropped. It's a check for pseudo keys.

Have "no" in the name confuses the fact that "yes" we are doing the check.

File: onyo/lib/onyo.py [hunk 2]
```diff
@@ -510,6 +512,32 @@ def _fsck_clean_tree(self) -> bool:
...
+        """
+        assets_with_pseudo_keys = {}
+        pseudo_keys = ["type", "make", "model", "serial"]
+
+        for asset in self.assets:
+            contents = self._read_asset(asset)
+
+            for p in pseudo_keys:
+                if p in contents:
+                    if asset not in assets_with_pseudo_keys.keys():
```
Comment (aqw) [inline onyo/lib/onyo.py:527]: No need for `.keys()` here. Simply: `if asset not in assets_with_pseudo_key:`

Nit: I find `assets_with_pseudo_keys.keys()` to be odd to read. Perhaps a more generic name, such as `assets_failed` or `assets_violation`?

File: onyo/lib/onyo.py [hunk 3]
```diff
@@ -510,6 +512,32 @@ def _fsck_clean_tree(self) -> bool:
...
         return True
 
+    def _fsck_no_pseudo_keys(self) -> bool:
+        """
+        Check that no asset contains any pseudo-key names.
+        """
+        assets_with_pseudo_keys = {}
+        pseudo_keys = ["type", "make", "model", "serial"]
+
+        for asset in self.assets:
```
Comment (aqw) [inline onyo/lib/onyo.py:522]: The rest of the for-loop can be simplified to:
```
violation_list = [x for x in pseudo_keys if x in self._read_asset(asset)]
if violation_list:
    assets_with_pseudo_keys[asset] = violation_list
```

File: onyo/lib/onyo.py [hunk 4]
```diff
@@ -510,6 +512,32 @@ def _fsck_clean_tree(self) -> bool:
...
+            contents = self._read_asset(asset)
+
+            for p in pseudo_keys:
+                if p in contents:
+                    if asset not in assets_with_pseudo_keys.keys():
+                        assets_with_pseudo_keys[asset] = []
+                    assets_with_pseudo_keys[asset].append(p)
+
+        if assets_with_pseudo_keys:
+            log.error('The following assets contain pseudo keys:\n' +
```
Comment (aqw) [inline onyo/lib/onyo.py:532]: This should be even more explicit. The user may not know what a pseudo key is, or that it is not allowed. Currently this message just states a fact. At minimum it should add "which are reserved for use in the asset filename."

File: onyo/lib/onyo.py [hunk 5]
```diff
@@ -510,6 +512,32 @@ def _fsck_clean_tree(self) -> bool:
...
+                      self._n_join([f'{asset}: ' +
+                                    ', '.join([str(k) for k in assets_with_pseudo_keys[asset]])
```
Comment (aqw) [inline onyo/lib/onyo.py:534]: `k` is already a string; No need to re-cast.

`assets_with_pseudo_keys[asset]` is a list not a dict.

`[str(k) for k in assets_with_pseudo_keys[asset]]` can simply be `assets_with_pseudo_keys[asset]`.

File: onyo/lib/onyo.py [hunk 6]
```diff
@@ -510,6 +512,32 @@ def _fsck_clean_tree(self) -> bool:
...
+                                    for asset in list(assets_with_pseudo_keys.keys())]))
```
Comment (aqw) [inline onyo/lib/onyo.py:535]: `.keys()` is not necessary to loop over a dict. And if it were, `.keys()` would not need to be cast into a list in order to iterate over it.

Needless casting is a common error across these PRs. Please only do so when you know that it is necessary.

File: tests/lib/test_Repo_fsck.py [hunk 7]
```diff
@@ -151,6 +150,33 @@ def test_fsck_yaml_invalid(caplog, repo):
         assert i in caplog.text
 
 
+contents = ["type: value",
```
Comment (aqw) [inline tests/lib/test_Repo_fsck.py:153]: Please also add a test where the "reserved" pseudo key is in the value, and thus should not fail. For example:

```
key: serial
```
and the more perverse:
```
key: "model: value"
```

**Review by aqw — CHANGES_REQUESTED** [review]

**Comment by TobiasKadelka** [comment]
I added the feedback (except the one I commented on) and pushed again, thanks.
I especially liked the other test-group of things that should still be allowed!

File: onyo/lib/onyo.py (hunk 3 above)
Comment (aqw) [reply onyo/lib/onyo.py:522]: Not that this looks to be a costly part of the code, but the list comprehension also reduces the number of assignments and appends dramatically.

**Comment by TobiasKadelka** [comment]
Well, I added/updated it now anyways. Not sure if the difference is relevant, but until we have performance checks for large numbers of assets, there is also no harm in optimizing a little bit. Thanks!

**Review by aqw — APPROVED** [review]
LGTM!

**Pull request merged**